File: benchmarks/latency_bench.cpp
#include "lfqueue/mpmc_queue.hpp"
#include "lfqueue/mutex_queue.hpp"

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <thread>
#include <vector>

// Payload carries the nanosecond timestamp of when it was pushed so that the
// consumer can measure end-to-end queue latency for each item.
struct Item {
    int64_t push_ns;
};
static_assert(std::is_nothrow_move_constructible<Item>::value, "");
static_assert(std::is_nothrow_move_assignable<Item>::value, "");

static int64_t now_ns() {
    return std::chrono::duration_cast<std::chrono::nanoseconds>(
               std::chrono::steady_clock::now().time_since_epoch())
        .count();
}

template <typename Queue>
void run_latency(const char* label, int n_samples, int queue_capacity) {
    Queue queue(static_cast<std::size_t>(queue_capacity));
    std::vector<int64_t> latencies;
    latencies.reserve(static_cast<std::size_t>(n_samples));

    std::atomic<int> consumed{0};

    std::thread consumer([&]() {
        Item item{0};
        while (consumed.load(std::memory_order_relaxed) < n_samples) {
            if (queue.pop(item)) {
                int64_t lat = now_ns() - item.push_ns;
                latencies.push_back(lat > 0 ? lat : 0);
                consumed.fetch_add(1, std::memory_order_relaxed);
            }
        }
    });

    for (int i = 0; i < n_samples; ++i)
        while (!queue.push(Item{now_ns()})) {}

    consumer.join();

    std::sort(latencies.begin(), latencies.end());
    auto pct = [&](double p) -> int64_t {
        std::size_t idx = static_cast<std::size_t>(
            static_cast<double>(latencies.size()) * p);
        if (idx >= latencies.size()) idx = latencies.size() - 1;
        return latencies[idx];
    };

    std::printf("%s,%lld,%lld,%lld\n",
                label,
                static_cast<long long>(pct(0.50)),
                static_cast<long long>(pct(0.99)),
                static_cast<long long>(pct(0.999)));
}

int main() {
    constexpr int kSamples  = 100'000;
    constexpr int kCapacity = 64;

    std::puts("queue,p50_ns,p99_ns,p999_ns");
    run_latency<lfqueue::MPMCQueue<Item>>("lfqueue", kSamples, kCapacity);
    run_latency<lfqueue::MutexQueue<Item>>("mutex",   kSamples, kCapacity);
    return 0;
}

File: benchmarks/latency_bench.rs
use oxide_broker::{BoundedQueue, BrokerEngine, MpmcQueue, MutexQueue};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::thread;
use std::time::Instant;

/// Payload carries the nanosecond timestamp of when it was pushed so that
/// the consumer can measure end-to-end queue latency for each item.
#[derive(Clone, Copy)]
struct Item {
    push_ns: i64,
}

fn now_ns(epoch: Instant) -> i64 {
    epoch.elapsed().as_nanos() as i64
}

fn run_latency<Q>(label: &str, n_samples: usize, queue_capacity: usize)
where
    Q: BoundedQueue<Item> + Send + Sync,
{
    let queue = Q::with_capacity(queue_capacity);
    let epoch = Instant::now();
    let consumed = AtomicUsize::new(0);

    let latencies = thread::scope(|scope| {
        let queue_ref = &queue;
        let consumed_ref = &consumed;
        let consumer = scope.spawn(move || {
            let mut latencies = Vec::with_capacity(n_samples);
            while consumed_ref.load(Ordering::Relaxed) < n_samples {
                if let Some(item) = queue_ref.pop() {
                    let lat = now_ns(epoch) - item.push_ns;
                    latencies.push(if lat > 0 { lat } else { 0 });
                    consumed_ref.fetch_add(1, Ordering::Relaxed);
                }
            }
            latencies
        });

        for _ in 0..n_samples {
            let mut item = Item { push_ns: now_ns(epoch) };
            while let Err(returned) = queue.push(item) {
                item = returned;
                item.push_ns = now_ns(epoch);
            }
        }

        consumer.join().unwrap()
    });

    let mut latencies = latencies;
    latencies.sort_unstable();
    let pct = |p: f64| -> i64 {
        let mut idx = (latencies.len() as f64 * p) as usize;
        if idx >= latencies.len() {
            idx = latencies.len() - 1;
        }
        latencies[idx]
    };

    println!("{},{},{},{}", label, pct(0.50), pct(0.99), pct(0.999));
}

/// Measures end-to-end pub/sub latency: time from `Publisher::publish` to
/// `Subscriber::try_recv` returning `Some(item)` on a 1-producer/1-subscriber
/// topology.
fn run_pubsub_latency(label: &str, n_samples: usize, queue_capacity: usize) {
    let broker = BrokerEngine::<Item>::new();
    broker.create_topic("bench", queue_capacity);

    let publisher = broker.publisher("bench").unwrap();
    let subscriber = broker.subscribe("bench").unwrap();
    let epoch = Instant::now();
    let consumed = AtomicUsize::new(0);

    let latencies = thread::scope(|scope| {
        let subscriber = &subscriber;
        let consumed_ref = &consumed;
        let consumer = scope.spawn(move || {
            let mut latencies = Vec::with_capacity(n_samples);
            while consumed_ref.load(Ordering::Relaxed) < n_samples {
                if let Some(item) = subscriber.try_recv() {
                    let lat = now_ns(epoch) - item.push_ns;
                    latencies.push(if lat > 0 { lat } else { 0 });
                    consumed_ref.fetch_add(1, Ordering::Relaxed);
                }
            }
            latencies
        });

        for _ in 0..n_samples {
            let mut item = Item { push_ns: now_ns(epoch) };
            loop {
                match publisher.publish(item) {
                    Ok(()) => break,
                    Err(_) => item = Item { push_ns: now_ns(epoch) },
                }
            }
        }

        consumer.join().unwrap()
    });

    let mut latencies = latencies;
    latencies.sort_unstable();
    let pct = |p: f64| -> i64 {
        let idx = ((latencies.len() as f64 * p) as usize).min(latencies.len() - 1);
        latencies[idx]
    };

    println!("{},{},{},{}", label, pct(0.50), pct(0.99), pct(0.999));
}

fn main() {
    const SAMPLES: usize = 100_000;
    const CAPACITY: usize = 64;

    println!("queue,p50_ns,p99_ns,p999_ns");
    run_latency::<MpmcQueue<Item>>("mpmc_raw", SAMPLES, CAPACITY);
    run_latency::<MutexQueue<Item>>("mutex_raw", SAMPLES, CAPACITY);

    // ── Pub/Sub end-to-end latency ──────────────────────────────────────
    // Measures the round-trip time from Publisher::publish() to
    // Subscriber::try_recv() returning Some(item).
    println!();
    println!("queue,p50_ns,p99_ns,p999_ns");
    run_pubsub_latency("oxide_broker", SAMPLES, CAPACITY);
}
